tests/trainer/test_fsdp_checkpoint.py

```diff
@@ -461,13 +467,76 @@ def test_fsdp_load_old_checkpoint(
     state_dict2 = trainer.state.state_dict()
 
     if (dist.get_global_rank() == 0 and state_dict_type == 'full') or state_dict_type == 'sharded':
-        filled_load_path = load_path.format(rank=dist.get_global_rank())
-        destination = str(tmp_path / pathlib.Path(filled_load_path).name)
-        get_file(filled_load_path, destination=destination)
-        with open(destination, 'rb') as f:
-            state_dict1 = torch.load(f)['state']
-        _compare_model_params_between_state_dicts(state_dict1, state_dict2)
+        if state_dict_type == 'sharded' and composer_version >= '0.16.0':
+            trainer2 = get_trainer(
+                num_features=32,  # This parameter setting is very important. Don't change or the test will fail.
+                num_classes=8,  # This parameter setting is very important. Don't change or the test will fail.
+                precision=precision,
+                max_duration='4ba',
+                train_metrics=train_metrics,
+                val_metrics=val_metrics,
+                fsdp_config=fsdp_config,
+            )
 
+            from torch.distributed import checkpoint as dist_cp
+
+            _, _, parsed_load_path = parse_uri(load_path)
+            gathered_tmp_path = str(dist.all_gather_object(tmp_path)[0])
+            destination = str(pathlib.Path(gathered_tmp_path) / parsed_load_path)
+
+            from composer.utils.checkpoint import DistCPObjectStoreReader
+            state_dict: Dict[str, Any] = {
+                'state': trainer2.state.state_dict(),
+                'rng': get_rng_state(),
+            }
+            if version.parse(torch.__version__) < version.parse('2.2.9'):
+                state_dict['state'].pop('optimizers')
+
+            object_store = S3ObjectStore(bucket=f'{s3_bucket}')
+            storage_reader = DistCPObjectStoreReader(source_path=parsed_load_path,
+                                                     destination_path=destination,
+                                                     object_store=object_store)
+
+            process_group = None
+            dist_cp.load_state_dict(
+                state_dict=state_dict,
+                storage_reader=storage_reader,
+                planner=None,
+                process_group=process_group,
+            )
+            if version.parse(torch.__version__) < version.parse('2.2.9'):
+                from torch.distributed.checkpoint.optimizer import load_sharded_optimizer_state_dict
+                from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
```

Comment:
Can we just call `load_sharded_checkpoint instead of copying logic? Especially because @eracah  pointed out this code has some issues and will be fixed soon? otherwise this will code rot

Reply:
@eracah we made some fixes to fsdp_load_checkpoint but I am not the expert in this code path so feel free to redesign the fix/test in a follow up PR.